src/aws_encryption_sdk/keyring/raw_keyring.py

```diff
@@ -0,0 +1,221 @@
+import struct
+
+import attr
+import six
+
+import aws_encryption_sdk.internal.formatting.deserialize
+import aws_encryption_sdk.internal.formatting.serialize
+from aws_encryption_sdk.exceptions import EncryptKeyError
+from aws_encryption_sdk.identifiers import WrappingAlgorithm
+from aws_encryption_sdk.internal.crypto.wrapping_keys import WrappingKey
+from aws_encryption_sdk.internal.str_ops import to_bytes
+from aws_encryption_sdk.keyring.base import Keyring
+from aws_encryption_sdk.structures import DataKey, MasterKeyInfo, RawDataKey
+
+
+def on_encrypt_helper(encryption_materials, key_provider, wrapping_key, wrapping_algorithm, key_name):
+    # Check if data key already exists
+    if not encryption_materials.data_encryption_key:
+
+        # Generate data key
+        plaintext_data_key = os.urandom(encryption_materials.algorithm.kdf_input_len)
+
+        # Check if data key is generated
+        if not plaintext_data_key:
+            return EncryptKeyError("Unable to generate data encryption key.")
+
+        # plaintext_data_key to RawDataKey
+        data_encryption_key = RawDataKey(key_provider=key_provider, data_key=plaintext_data_key)
+
+        # Add generated data key to encryption_materials
+        encryption_materials.add_data_encryption_key(data_encryption_key, encryption_materials.keyring_trace)
+
+    else:
+        plaintext_data_key = encryption_materials.data_encryption_key
+
+    # Encrypt data key
+    encrypted_wrapped_key = wrapping_key.encrypt(
+        plaintext_data_key=plaintext_data_key, encryption_context=encryption_materials.encryption_context
+    )
+
+    # EncryptedData to EncryptedDataKey
+    encrypted_data_key = aws_encryption_sdk.internal.formatting.serialize.serialize_wrapped_key(
+        key_provider=key_provider,
+        wrapping_algorithm=wrapping_algorithm,
+        wrapping_key_id=key_name,
+        encrypted_wrapped_key=encrypted_wrapped_key,
+    )
+
+    # Add encrypted data key to encryption_materials
+    encryption_materials.add_encrypted_data_key(encrypted_data_key, encryption_materials.keyring_trace)
+
+    return encryption_materials
+
+
+def on_decrypt_helper(decryption_materials, wrapping_key, wrapping_algorithm, key_name, encrypted_data_key):
+    # Check if plaintext data key exists
+    if decryption_materials.data_key:
+        return decryption_materials
+
+    # Wrapped EncryptedDataKey to deserialized EncryptedData
+    encrypted_wrapped_key = aws_encryption_sdk.internal.formatting.deserialize.deserialize_wrapped_key(
+        wrapping_algorithm=wrapping_algorithm, wrapping_key_id=key_name, wrapped_encrypted_key=encrypted_data_key
+    )
+
+    # EncryptedData to raw key string
+    plaintext_data_key = wrapping_key.decrypt(
+        encrypted_wrapped_data_key=encrypted_wrapped_key, encryption_context=decryption_materials.encryption_context
+    )
+
+    # Update decryption materials
+    data_encryption_key = DataKey(
+        key_provider=encrypted_data_key.key_provider,
```

Comment:
This should be a new `MasterKeyInfo` created using this keyring. This is especially important for the `RawAESKeyring` since key name and provider info are different for that keyring.